### ISSUE TITLE
imploded arrays VS serialized arrays VS json_encoded arrays?

### ISSUE THREAD

**Issue by emanuele45** [elkarte/Elkarte#2489]
While testing out the upgrade process I was wondering: what about take the opportunity and finally standardize the approach to storing arrays in the database? (At least for the settings table.)

At the moment we have imploded strings and serialized strings.
The first are simple but limited (for example it's impossible to store the indexes of the elements of the array).
The second are a pain if you have to change the data manually (you have to change also the size of the entry).

So I was wondering if use json could be the "definitive" approach: it can store anything (arrays, hash, indexes, etc.), there is nothing to "rebuild" in case data are changed by hands, it's nicer to see in code (in certain places we have empty serialized arrays and are somehow ugly).

Opinions?
If we decide to do that, we could even have it done directly in updateSettings (is_array => json_encode), not sure about reloadSettings instead.

**Comment by Vekseid** [comment]
Serialize breaks when you look at it funny. It does not play nice with Redis.

Implode is fine for simple stuff, but json is what I used in my personal project.

**Comment by Frenzie** [comment]
In principle JSON is just an alternative serialization format, but besides
being easier to edit another advantage is that you'd be able to pass it
straight to JS if desired instead of having to jump through extra hoops. On
the downside, imploded strings are significantly easier still, but that's
probably not worth any extra code complexity.

PS I don't really understand what you mean by rebuild?

**Comment by Vekseid** [comment]
Yeah piping to JS is enormously handy. It's best to create a small wrapper function to set the constants needed to make php and js speak the same json.

**Comment by albertlast** [comment]
Maybe some interesting information from postgresql side:
postgresql is able for every datatype to handle this as array: http://www.postgresql.org/docs/9.1/static/arrays.html since ages
since pg 9.2 exists the datatype json which is internaly a text field with some addationaly opertators: http://www.postgresql.org/docs/9.4/static/functions-json.html
since pg 9.4 exists jsonb which is more advance version

**Comment by emanuele45** [comment]
Interesting, interesting!
At the moment I don't remember any column that only contains json data, usually serialized/imploded data are stored alongside plain text data, so mixed stuff.
I wonder if some particular tables could benefit from storing data in a json format, for example the themes table comes to my mind, several options normally stored as id_member/key/value ternaries could be "compacted" into single entries id_member/json-array (especially considering in most of the cases we fetch the whole bunch of stuff and use it as such rather than search for a particular key, it seems (in PSQL, not sure in MySQL) possible to update single elements of the array, so even change a particular value for all the members should be not different from what we do now).

Of course we'd have to have a mysql counterpart, otherwise maintain two different ways of storing data would be a pain... and apparently only mysql 5.7 has native support for json data... :disappointed:

**Comment by albertlast** [comment]
> I wonder if some particular tables could benefit from storing data in a json format, for example the themes table comes to my mind, several options normally stored as id_member/key/value ternaries could be "compacted" into single entries id_member/json-array (especially considering in most of the cases we fetch the whole bunch of stuff and use it as such rather than search for a particular key, it seems (in PSQL, not sure in MySQL) possible to update single elements of the array, so even change a particular value for all the members should be not different from what we do now).

I got two possible reason for this topic:
- to use tages http://www.databasesoup.com/2015/01/tag-all-things.html
- custom columns for thirdparty solution 
  every table could get a json column, so when a thirdparty needs a new "column" they only add keys value to the existing json -> don't need to extend the schema

**Comment by h-collector** [comment]
Just to add my 2 cents.
Migration (or synchronization) to another server with simple `sed` (changing paths) on database dump:
`$modSettings['attachmentUploadDir'] = unserialize($modSettings['attachmentUploadDir']);`
`=> Undefined index: basedirectory_for_attachments`
with json encoded settings I wouldn't need to fix it by hand
